lib/Sema/TypeCheckDeclPrimary.cpp

```diff
@@ -1401,6 +1402,188 @@ static void maybeDiagnoseClassWithoutInitializers(ClassDecl *classDecl) {
   diagnoseClassWithoutInitializers(classDecl);
 }
 
+static bool isModuleQualified(TypeRepr *repr, ModuleDecl *module) {
+  auto extIdent = dyn_cast_or_null<CompoundIdentTypeRepr>(repr);
+  if (!extIdent) {
+    return false;
+  }
+  
+  auto components = extIdent->getComponents();
+  if (components.size() == 1) {
+    return false;
+  }
+  
+  return components.front()->getNameRef().isSimpleName(module->getName());
+}
+
+static void diagnoseRetroactiveConformances(ExtensionDecl *ext) {
+  ModuleDecl *module = ext->getParentModule();
+
+  // We only want to warn about this if a module has library evolution enabled,
+  // which we assume means it's a library meant to be distributed to clients.
+  if (!module->isResilient()) {
+    return;
+  }
+
+  // Disable this for the Foundation module because of the interplay with
+  // _ObjectiveCBridgeable.
+  if (module->isFoundationModule()) {
+    return;
+  }
+
+  // Don't warn for this if we see it in module interfaces (that usually
+  // indicates the interface was built with
+  // -module-interface-preserve-types-as-written)
+  if (ext->getParentSourceFile()->Kind == SourceFileKind::Interface) {
+    return;
+  }
+
+  Type extendedType = ext->getExtendedType();
+  NominalTypeDecl *extendedNominalDecl = ext->getExtendedNominal();
+  if (!extendedNominalDecl) {
+    return;
+  }
+
+  // If it's a type in the same module, conformances aren't retroactive.
+  ModuleDecl *extTypeModule = extendedNominalDecl->getParentModule();
+  if (extTypeModule == module) {
+    return;
+  }
+
+  // If we're compiling an overlay and the type comes from the clang module,
+  // conformances aren't retroactive.
+  if (auto clangModule = extTypeModule->findUnderlyingClangModule()) {
+    if (clangModule->getTopLevelModuleName() == module->getName().str()) {
+      return;
+    }
+  }
+
+  // At this point, we know we're extending a type declared outside this module.
+  // We better only be conforming it to protocols declared within this module.
+  llvm::SmallVector<ProtocolDecl *, 8> workList;
+  for (TypeLoc typeLoc : ext->getInherited()) {
+    if (isModuleQualified(typeLoc.getTypeRepr(), module)) {
+      continue;
+    }
+    auto protoDecl = 
+        dyn_cast_or_null<ProtocolDecl>(typeLoc.getType()->getAnyNominal());
+    if (!protoDecl) {
+      continue;
+    }
+    workList.push_back(protoDecl);
+  }
+
+  // Search through the whole inheritance tree to find protocols that this
+  // extension declares a conformance to.
+
+  llvm::SmallPtrSet<ProtocolDecl *, 8> seen;
+  llvm::SmallSetVector<ProtocolDecl *, 8> externalProtocols;
+  while (!workList.empty()) {
+    auto protoDecl = workList.pop_back_val();
+
+    // First, push the inherited protocols onto the worklist.
+    for (ProtocolDecl *inherited : protoDecl->getInheritedProtocols()) {
+      if (seen.insert(inherited).second) {
+        workList.push_back(inherited);
```

Comment:
A slight optimization of this could be to break the loop after you've encountered a protocol you've already seen, since you know that its ancestors have all been pushed onto the worklist by this very loop.